app/services/slack_service.rb

```diff
@@ -0,0 +1,200 @@
+
+    user = User.where(:slack_registration_token => slack_register_token).take
+
+    raise InvalidCommand.new 'Invalid registration token' unless user != nil
+
+    raise InvalidCommand.new 'This kudo-o-matic account is already linked to Slack.' unless user.slack_id == nil
+
+    user.slack_id = user_id
+    user.slack_registration_token = nil
+    raise InvalidCommand.new "That didn't quite work, #{user.errors.full_messages.join(', ')}" unless user.save
+  end
+
+  def self.add_to_workspace(code, team_id)
+    client = Slack::Web::Client.new
+    if code == nil
+      raise InvalidRequest.new 'Auth token is missing'
+    end
+
+    if team_id == nil
+      raise InvalidRequest.new 'Team id is missing'
+    end
+    team = Team.find(team_id)
+
+    rc = client.oauth_v2_access(
+        code: code,
+        client_id: ENV['SLACK_CLIENT_ID'],
+        client_secret: ENV['SLACK_CLIENT_SECRET']
+    )
+
+    team.channel_id = rc['incoming_webhook']['channel_id']
+    team.slack_bot_access_token = rc["access_token"]
+    team.slack_team_id = rc["team"]["id"]
+
+    raise InvalidRequest.new "That didn't quite work, #{team.errors.full_messages.join(', ')}" unless team.save
+
+    send_welcome_message(team)
+  end
+
+  def self.list_guidelines(slack_team_id)
+    team = Team.find_by_slack_team_id(slack_team_id)
+
+    raise InvalidCommand.new 'No team with that Slack ID' unless team != nil
+
+    guidelines = Guideline.where(:team => team).order(:kudos)
+
+    if guidelines.count == 0
+      return create_markdown_block("No guidelines")
+    else
+      return create_markdown_block(guidelines_to_list(guidelines))
+    end
+  end
+
+  def self.get_oauth_url(team_id)
+    URI::HTTP.build(
+        :host => Settings.slack_auth_endpoint,
+        :path => '/oauth/v2/authorize',
+        :query => {
+            :state => team_id,
+            :client_id => ENV['SLACK_CLIENT_ID'],
+            :scope => Settings.slack_scopes,
+            :user_scope => Settings.slack_user_scopes
+        }.to_query
+    )
+  end
+
+  private
+
+  def self.send_welcome_message(team)
+    client = Slack::Web::Client.new(token: team.slack_bot_access_token)
+
+    client.chat_postMessage(channel: team.channel_id, text: "Is it a bird? is it a plane? It's Kudo-O-Matic!")
+  end
+
+  def self.get_post_receivers(post)
+    receiver_string = ""
+    post.receivers.each_with_index do |receiver, index|
+      receiver_string += receiver.slack_id == nil ? "#{receiver.name}" : "<@#{receiver.slack_id}>"
+
+      if post.receivers.count > 1 && index != post.receivers.count - 1
+        receiver_string += (index == (post.receivers.count - 2)) ? ' and ' : ', '
+      end
+    end
+
+    receiver_string
+  end
+
+  def self.get_receivers_from_command(text)
+    slack_users = text.scan(/(?=<).*?(?<=>)/)
+
+    receivers = []
+
+    slack_users.each do |slack_user|
+      id = slack_user[slack_user.index('@') + 1..slack_user.index('|') - 1]
```

Comment:
Can't the Ruby Slack client handle this kind of parsing for you?

Reply:
As far as I can find it can't It can only [escape messages](https://github.com/slack-ruby/slack-ruby-client/blob/master/README.md#message-parsing) to show usernames or user id's but it doesn't offer any methods to extract the id's.

Reply:
Ok, bummer!